portiaui/app/utils/selectors.js

```diff
@@ -879,6 +879,71 @@ export function makeItemsFromGroups(groups) {
     return items;
 }
 
+function createSelectorGenerators(structure, selectorMatcher, accumulator) {
+    const generators = [];
+
+    for (let element of structure) {
+        const {annotation, children} = element;
+        let selectorGenerator;
+        if (children) {
+            selectorGenerator = ContainerSelectorGenerator.create({});
+            selectorGenerator.addChildren(
+                createSelectorGenerators(children, selectorMatcher, accumulator));
+
+        } else {
+            selectorGenerator = AnnotationSelectorGenerator.create({
+                selectorMatcher,
+                annotation
+            });
+        }
+        generators.push(selectorGenerator);
+        accumulator.push([annotation, selectorGenerator]);
+    }
+
+    return generators;
+}
+
+export function updateStructureSelectors(structure, selectorMatcher) {
+    const accumulator = [];
+    createSelectorGenerators(structure, selectorMatcher, accumulator);
+    for (let [annotation, selectorGenerator] of accumulator) {
+        const selector = selectorGenerator.get('selector');
+        if (selectorGenerator instanceof AnnotationSelectorGenerator) {
```

Comment:
Can you replace these conditionals with polymorphism? Have the generator classes respond to a method which takes care of updating the structure given the annotation. In case the selector generator is not a `AnnotationSelectorGenerator` or `ContainerSelectorGenerator`, just do `this.destroy()`.